File: backend/app/schemas/booking_reminder.py
import datetime as dt
import uuid
from typing import Literal

from pydantic import BaseModel, ConfigDict

ReminderStatus = Literal["pending", "sent", "cancelled", "expired"]

# Allowed offsets in minutes (prevent abuse)
ALLOWED_OFFSETS = {15, 30, 60, 120, 180, 360, 720, 1440, 2880}
ALLOWED_CHANNELS = {"email", "push", "telegram"}
MAX_REMINDERS_PER_BOOKING = 5


class BookingReminderCreate(BaseModel):
    """User provides the offset and channels; remind_at is computed server-side."""

    offset_minutes: int
    channels: list[str] = ["push"]


class BookingReminderRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    booking_id: uuid.UUID
    offset_minutes: int
    channels: list[str]
    remind_at: dt.datetime
    status: ReminderStatus
    created_at: dt.datetime


class BookingReminderListResponse(BaseModel):
    items: list[BookingReminderRead]


class ReminderOffsetEntry(BaseModel):
    """A single default reminder: offset + which channels to use."""

    offset_minutes: int
    channels: list[str] = ["push"]


class DefaultReminderOffsetsRead(BaseModel):
    default_reminder_offsets: list[ReminderOffsetEntry]


class DefaultReminderOffsetsUpdate(BaseModel):
    default_reminder_offsets: list[ReminderOffsetEntry]

File: frontend/src/stores/bookingReminder.ts
import { ref } from 'vue'

import { defineStore } from 'pinia'

import { useAuthenticatedClient } from '@/composables/useAuthenticatedClient'

export interface BookingReminder {
  id: string
  booking_id: string
  offset_minutes: number
  channels: string[]
  remind_at: string
  status: 'pending' | 'sent' | 'cancelled' | 'expired'
  created_at: string
}

export interface ReminderOffsetEntry {
  offset_minutes: number
  channels: string[]
}

export interface DefaultReminderOffsets {
  default_reminder_offsets: ReminderOffsetEntry[]
}

export const ALLOWED_OFFSETS = [15, 30, 60, 120, 180, 360, 720, 1440, 2880] as const
export const ALLOWED_CHANNELS = ['email', 'push', 'telegram'] as const

export const useBookingReminderStore = defineStore('bookingReminder', () => {
  const { get, post, put, delete: del } = useAuthenticatedClient()

  const defaultOffsets = ref<ReminderOffsetEntry[]>([])

  // ── Default reminder offsets ────────────────────────────────────

  async function fetchDefaultOffsets(): Promise<ReminderOffsetEntry[]> {
    const res = await get<{ data: DefaultReminderOffsets }>({
      url: '/users/me/reminder-defaults',
    })
    defaultOffsets.value = res.data.default_reminder_offsets
    return defaultOffsets.value
  }

  async function updateDefaultOffsets(
    offsets: ReminderOffsetEntry[],
  ): Promise<ReminderOffsetEntry[]> {
    const res = await put<{ data: DefaultReminderOffsets }>({
      url: '/users/me/reminder-defaults',
      body: { default_reminder_offsets: offsets },
    })
    defaultOffsets.value = res.data.default_reminder_offsets
    return defaultOffsets.value
  }

  // ── Per-booking reminders ───────────────────────────────────────

  async function fetchBookingReminders(bookingId: string): Promise<BookingReminder[]> {
    const res = await get<{ data: { items: BookingReminder[] } }>({
      url: `/bookings/${bookingId}/reminders`,
    })
    return res.data.items
  }

  async function addBookingReminder(
    bookingId: string,
    offsetMinutes: number,
    channels: string[],
  ): Promise<BookingReminder> {
    const res = await post<{ data: BookingReminder }>({
      url: `/bookings/${bookingId}/reminders`,
      body: { offset_minutes: offsetMinutes, channels },
    })
    return res.data
  }

  async function deleteReminder(reminderId: string): Promise<void> {
    await del({ url: `/reminders/${reminderId}` })
  }

  return {
    defaultOffsets,
    fetchDefaultOffsets,
    updateDefaultOffsets,
    fetchBookingReminders,
    addBookingReminder,
    deleteReminder,
  }
})
